src/types/any.rs

```diff
@@ -1210,35 +1189,35 @@ impl<'py> PyAnyMethods<'py> for Bound<'py, PyAny> {
         N: IntoPy<Py<PyString>>,
         A: IntoPy<Py<PyTuple>>,
     {
-        self.getattr(name)
-            .and_then(|method| method.call(args, kwargs))
+        // Don't `args.into_py()`! This will lose the optimization of vectorcall.
+        match kwargs {
+            Some(_) => self
+                .getattr(name)
+                .and_then(|method| method.call(args, kwargs)),
+            None => self.call_method1(name, args),
+        }
     }
 
+    #[inline]
     fn call_method0<N>(&self, name: N) -> PyResult<Bound<'py, PyAny>>
     where
         N: IntoPy<Py<PyString>>,
     {
-        cfg_if::cfg_if! {
-            if #[cfg(all(Py_3_9, not(any(Py_LIMITED_API, PyPy, GraalPy))))] {
-                let py = self.py();
-
-                // Optimized path on python 3.9+
-                unsafe {
-                    let name = name.into_py(py).into_bound(py);
-                    ffi::PyObject_CallMethodNoArgs(self.as_ptr(), name.as_ptr()).assume_owned_or_err(py)
-                }
-            } else {
-                self.call_method(name, (), None)
-            }
-        }
+        // This will use optimized call (if available), as per `<() as IntoPy<Py<PyTuple>::__py_call_method_vectorcall1()`.
+        self.call_method1(name, ())
```

Comment:
Maybe again a case where if we had `ffi::compat::PyObject_CallMethodNoArgs` we could just use that here and it'd be more obvious?

Reply:
How can I define `ffi::compat::PyObject_CallMethodNoArgs`, considering that `Bound` isn't available in `pyo3-ffi` (for `getattr()`) and without complicating it needlessly?

Reply:
I will take a quick look at this later and judge how hard it is...

Reply:
#4461 looks like this is trivial; I'll merge that and then you can rebase :)